server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -53,16 +66,57 @@ public CompletableFuture<IndexInput> asyncFetchBlob(Path path, Supplier<IndexInp
         return invocationLinearizer.linearize(path, p -> indexInputSupplier.get());
     }
 
+    /*
+    This method accessed through the ConcurrentInvocationLinearizer so read-check-write is acceptable here
+     */
     private IndexInput fetchBlob(BlobFetchRequest blobFetchRequest) throws IOException {
-        // for first phase, this is a simple remote repo blob read with no caching at all
+        // check if the origin is already in block cache
+        IndexInput origin = fileCache.computeIfPresent(blobFetchRequest.getFilePath(), (path, cachedIndexInput) -> {
+            if (cachedIndexInput.isClosed()) {
+                // if it's already in the file cache, but closed, open it and replace the original one
+                try {
+                    IndexInput luceneIndexInput = blobFetchRequest.getDirectory().openInput(blobFetchRequest.getFileName(), IOContext.READ);
+                    return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), luceneIndexInput);
+                } catch (IOException ioe) {
+                    logger.warn("Open index input " + blobFetchRequest.getFilePath() + " got error ", ioe);
+                    // open failed so return null to download the file again
+                    return null;
+                }
+
+            }
+            // already in the cache and ready to be used (open)
+            return cachedIndexInput;
+        });
+
+        if (Objects.isNull(origin)) {
+            // origin is not in file cache, download origin
+
+            // open new origin
+            IndexInput downloaded = downloadBlockLocally(blobFetchRequest);
+
+            // refcount = 0 at the beginning
+            FileCachedIndexInput newOrigin = new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), downloaded);
+
+            // put origin into file cache
+            fileCache.put(blobFetchRequest.getFilePath(), newOrigin);
+            origin = newOrigin;
+        }
+        // always, need to clone to do refcount += 1, and rely on GC to clean these IndexInput which will refcount -= 1
```

Comment:
See my comment here: https://github.com/opensearch-project/OpenSearch/issues/5243#issuecomment-1370090311

I may be misunderstanding things, but it seems to me that it should be possible to implement this such that we wouldn't need to rely on a GC cleaner to decrement refcounts.

Reply:
as we discussed, relying on GC seems to be necessary for now